{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import defaultdict\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "REFSEQ_DIR = \"RefSeq\"\n",
    "\n",
    "PRODUCT_NAME_MAPPING = {\n",
    "    \"leader protein\": \"nsp1\",\n",
    "    \"3C-like proteinase\": \"nsp5\",\n",
    "    \"RNA-dependent RNA polymerase\": \"nsp12\",\n",
    "    \"helicase\": \"nsp13\",\n",
    "    \"3'-to-5' exonuclease\": \"nsp14\",\n",
    "    \"endoRNAse\": \"nsp15\",\n",
    "    \"2'-O-ribose methyltransferase\": \"nsp16\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "annoRef = SeqIO.read(os.path.join(REFSEQ_DIR, \"SARS_CoV_2.gb\"), \"genbank\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF1ab\n",
      "S\n",
      "ORF3a\n",
      "E\n",
      "M\n",
      "ORF6\n",
      "ORF7a\n",
      "ORF7b\n",
      "ORF8\n",
      "N\n",
      "ORF10\n"
     ]
    }
   ],
   "source": [
    "aa2nucleotide = {}\n",
    "mat2nucleotide = {}\n",
    "matAA2nucleotide = {}\n",
    "proteinSeqs = {}\n",
    "orfNames = []\n",
    "\n",
    "peptideIndex = 0\n",
    "\n",
    "for f in annoRef.features:\n",
    "    if f.type == \"CDS\":\n",
    "        geneName = f.qualifiers['gene'][0]\n",
    "        if geneName not in orfNames:\n",
    "            orfNames.append(geneName)\n",
    "            aaSeq = f.qualifiers['translation'][0]\n",
    "            proteinSeqs[geneName] = aaSeq\n",
    "            print(geneName)\n",
    "\n",
    "            aaIndex = 0\n",
    "            codonPos = 1\n",
    "            codonIndices = []\n",
    "            siteMapping = {}\n",
    "            for i in f.location.parts:\n",
    "                for siteIndex in range(i.start, i.end):\n",
    "                    codonIndices.append(siteIndex + 1)\n",
    "                    codonPos += 1\n",
    "                    if codonPos > 3:\n",
    "                        aaIndex += 1\n",
    "                        peptideIndex += 1\n",
    "                        siteMapping[tuple(codonIndices)] = (aaIndex, peptideIndex)\n",
    "                        codonPos = 1\n",
    "                        codonIndices = []\n",
    "                aa2nucleotide[geneName] = siteMapping\n",
    "    elif f.type == 'mat_peptide':\n",
    "        productName = f.qualifiers['product'][0]\n",
    "        if productName == \"nsp11\":\n",
    "            continue\n",
    "\n",
    "        aaIndex = 0\n",
    "        codonPos = 1\n",
    "        codonIndices = []\n",
    "        siteMapping = {}\n",
    "        for i in f.location.parts:\n",
    "            for siteIndex in range(i.start, i.end):\n",
    "                codonIndices.append(siteIndex + 1)\n",
    "                codonPos += 1\n",
    "                if codonPos > 3:\n",
    "                    aaIndex += 1\n",
    "#                     peptideIndex += 1\n",
    "                    siteMapping[tuple(codonIndices)] = (aaIndex, peptideIndex)\n",
    "                    codonPos = 1\n",
    "                    codonIndices = []\n",
    "                if productName in PRODUCT_NAME_MAPPING:\n",
    "                    productName = PRODUCT_NAME_MAPPING[productName]\n",
    "                mat2nucleotide[siteIndex + 1] = productName\n",
    "            matAA2nucleotide[productName] = siteMapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aa2nucleotide = {}\n",
    "# mat2nucleotide = {}\n",
    "# proteinSeqs = {}\n",
    "# orfNames = []\n",
    "\n",
    "# peptideIndex = 0\n",
    "\n",
    "# for f in annoRef.features:\n",
    "#     if f.type == \"CDS\":\n",
    "#         geneName = f.qualifiers['gene'][0]\n",
    "#         if geneName not in orfNames:\n",
    "#             orfNames.append(geneName)\n",
    "#             aaSeq = f.qualifiers['translation'][0]\n",
    "#             proteinSeqs[geneName] = aaSeq\n",
    "#             print(geneName)\n",
    "\n",
    "#             aaIndex = 0\n",
    "#             codonPos = 1\n",
    "#             codonIndices = []\n",
    "#             siteMapping = {}\n",
    "#             for i in f.location.parts:\n",
    "#                 for siteIndex in range(i.start, i.end):\n",
    "#                     codonIndices.append(siteIndex + 1)\n",
    "#                     codonPos += 1\n",
    "#                     if codonPos > 3:\n",
    "#                         aaIndex += 1\n",
    "#                         peptideIndex += 1\n",
    "#                         siteMapping[tuple(codonIndices)] = (aaIndex, peptideIndex)\n",
    "#                         codonPos = 1\n",
    "#                         codonIndices = []\n",
    "#                 aa2nucleotide[geneName] = siteMapping\n",
    "#     elif f.type == 'mat_peptide':\n",
    "#         productName = f.qualifiers['product'][0]\n",
    "#         for i in f.location.parts:\n",
    "#             for siteIndex in range(i.start, i.end):\n",
    "#                 mat2nucleotide[siteIndex + 1] = productName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = []\n",
    "\n",
    "for geneName in aa2nucleotide:\n",
    "    aaSeq = proteinSeqs[geneName]\n",
    "    for codonIndices in aa2nucleotide[geneName]:\n",
    "        aaSite, peptideSite = aa2nucleotide[geneName][codonIndices]\n",
    "        if aaSite <= len(aaSeq):\n",
    "            aa = aaSeq[aaSite - 1]\n",
    "        else:\n",
    "            aa = '*'\n",
    "        for site in codonIndices:\n",
    "            if site in mat2nucleotide:\n",
    "                productName = mat2nucleotide[site]\n",
    "                aaSite = matAA2nucleotide[productName][codonIndices][0]\n",
    "            else:\n",
    "                productName = geneName\n",
    "            nucleotide = annoRef[site - 1]\n",
    "            df.append([geneName,\n",
    "                       productName,\n",
    "                       aaSite,\n",
    "                       peptideSite,\n",
    "                       aa,\n",
    "                       site,\n",
    "                       nucleotide])\n",
    "\n",
    "df = pd.DataFrame(df, columns=['gene',\n",
    "                               'product',\n",
    "                               'aaPos',\n",
    "                               'peptidePos',\n",
    "                               'aa',\n",
    "                               'genomePos',\n",
    "                               'nucleotide'])\n",
    "df.to_csv(os.path.join(REFSEQ_DIR, \"SARS_CoV_2.csv\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
